samba4: add package samba-4.8

Andy2244 opened the pull request:
Maintainer: me
Compile tested: arm/mips (master)
Run tested: arm/mvebu (master)

Description:
This is my samba 4.8 package, it includes the server with vfs plugin support and tools, with extended config options (fileserver, netbios, ad-dc, winbind, avahi). It also includes a WSD Daemon to replace netbios, so Win10  clients can see samba shares via explorer.
_The corresponding luci-ui package openwrt/luci#1985 was already merged._

Signed-off-by: Andy Walsh <andy.walsh44+github@gmail.com>

Andy2244 commented:
Just confirmed on a fresh compile test, target x86 fails because e2fsprogs is build in-between krb5 and samba4.
Someone send me patches for e2fsprogs/krb5 so libcom_err, libss are separate packages and can be reused by krb5, trying to test those now.

poranje commented:
Does this apply to the error table generated with compile_et only or should the libcom_err always be statically be linked with all programs ? I ask this because the table I is program specific, isn't that so ?
Paul


> Op 3 feb. 2018, om 18:06 heeft Andy Walsh <notifications@github.com> het volgende geschreven:
> 
> Just confirmed on a fresh compile test, target x86 fails because e2fsprogs is build in-between krb5 and samba4.
> Someone send me patches for e2fsprogs/krb5 so libcom_err, libss are separate packages and can be reused by krb5, trying to test those now.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Andy2244 commented:
@poranje with [#714](https://github.com/openwrt/openwrt/pull/714) and #5562 krb5 accepts the system com_err/ss and in return samba4 accepts krb5 libs. As far as i understand com_err can be system-wide shared, but packages generate some static code with a compatible compile_et. At least thats what krb5 seems to-do, it tests the system com_err lib and also the compile_et tool.

Andy2244 commented:
Updated to stable 4.8 branch and now passes travis checks, should still be merged alongside #5562.

lucize commented:
also tested and working on mvebu due to samba3.6 not compiling anymore on my system (maybe gcc7.3 problems can't say)
 
@hnyman can we have this series in, samba 3.6 is long EOL ?

Regards

Andy2244 commented:
@lucize I cant judge if its "needed", but the final bins are linked against all those libs so they are needed to run the selected targets/features. I use everything that could be done with the default waf logic and without heavily patching the source to minimize size. So i only select the targets that are required for each feature and made the features select-able, so the minimal size is a server only (no netbios/vfs/ad-dc) package at around 4.2-4.6 MB size.

If anyone finds a trick to minimize it more, i'm happy to add the patches as long as they are reasonable to maintain.

lucize commented:
ok, I'll test it one by one and come back with an answer, maybe with some ifdef based on build selection some libs can be excluded

Regards

Andy2244 commented:
@lucize You can try, but keep in mind that waf is already supposedly doing this. So if you tell waf to only build "smbstatus" than it will resolve all needed libs and only build/include those.

Andy2244 commented:
@diizzyy kinda, the existing lib packages are not using proper cross-compile answer files per architecture or qemu. So i doubt anyone actually tested the libs on all openwrt architectures. Thats why i use qemu in this package, since i did not want to guess cross-answers or update the files every few versions.

diizzyy commented on net/samba4/Config.in:
```diff
@@ -0,0 +1,111 @@
+
+config SAMBA4_SERVER
+	bool "Server (fileserver)"
+	select BUSYBOX_CONFIG_RENICE
```
This is a bad idea, don't modify other package "indirectly". Submit that it should be enabled by default upstream instead if it's really needed however my guess is that it's not the end of the world if renice isn't available.

Andy2244 replied:
Probably true, i think it was originally used to prevent the samba process itself to block other network related work, so it does not stall itself. I never tested it without renice, that's why i kept it in and found no other way to include it.
So there is no "official" way to have a package depend on a specific config option? How do other packages handle this, say for example if they depend on a kernel mod or specific kernel option?

diizzyy replied:
You can depend on kernel mods however there's no good way of depending on options that doesn't result in a unique package to my knowledge. @karlp @hnyman

diizzyy commented on net/samba4/Config.in:
```diff
@@ -0,0 +1,111 @@
+		help
+			provides commonly used vfs modules
+			installs: 
+				modules: vfs_btrfs vfs_fruit vfs_shadow_copy vfs_shadow_copy2 vfs_recycle vfs_fake_perms vfs_readonly vfs_cap vfs_offline vfs_crossrename
```
vfs_btrfs should depend on kmod-fs-btrfs?
From what I can tell you only need vfs_shadow_copy2 not vfs_shadow_copy.
https://github.com/samba-team/samba/blob/master/source3/modules/vfs_shadow_copy.c
https://github.com/samba-team/samba/blob/master/source3/modules/vfs_shadow_copy2.c

Andy2244 replied:
Yes for vfs_btrfs, will add a config check and i actually did not check for vfs_shadow_copy, but i think that was the samba default output via "smbd -b", so i kept it close to the default.

diizzyy commented on net/samba4/Config.in:
```diff
@@ -0,0 +1,111 @@
+		help
+			provides extra vfs modules
+			installs: 
+				modules: vfs_virusfilter vfs_shell_snap vfs_snapper vfs_commit vfs_worm vfs_xattr_tdb vfs_streams_xattr vfs_aio_fork vfs_aio_pthread vfs_linux_xfs_sgid vfs_netatalk vfs_dirsort vfs_fileid
```
vfs_snapper --> Requires snapper (not in repo)
vfs_aio_* --> Not sure we needs those at all as we have optional aio support?
vfs_linux_xfs_sgid --> should depend on kmod-fs-xfs?

diizzyy commented on net/samba4/Config.in:
```diff
@@ -0,0 +1,111 @@
+		default y
+	
+	config SAMBA4_SERVER_VFSX
+		bool "enable extendet VFS modules"
```
typo

diizzyy commented on net/samba4/Makefile:
```diff
@@ -0,0 +1,296 @@
+#
+# Copyright (C) 2007-2017 OpenWrt
```
Don't assign copyright to OpenWrt unless you're affiliated, use your own name or omit it.

diizzyy commented:
I see, I left some comments but it looks great otherwise. :-)

Andy2244 commented on net/samba4/Makefile:
```diff
@@ -0,0 +1,296 @@
+  DEPENDS:=+zlib +libpthread +libpopt +libcap +libtirpc +libnettle +krb5-libs \
+	+SAMBA4_SERVER_VFSX:libdbus +SAMBA4_SERVER_ACL:acl \
+	+SAMBA4_SERVER_AVAHI:libavahi-client +SAMBA4_SERVER_AVAHI:avahi-dbus-daemon \
+	+SAMBA4_SERVER_AD_DC:python-base +SAMBA4_SERVER_AD_DC:krb5-server +SAMBA4_SERVER_AD_DC:libopenssl +SAMBA4_SERVER_AD_DC:libgnutls +SAMBA4_SERVER_AD_DC:libopenldap
```
Yes all libs that are needed for each configuration was step by step tested.

Andy2244 commented:
@diizzyy ok, updated to 4.8.1 and done the suggested changes.

diizzyy commented:
Great, I think the qemu case is valid so it's fine as it is right now.
I'll give it a try in the weekend.

Andy2244 commented:
trivial update to samba 4.8.2 and qemu 2.12.0

Andy2244 commented:
* add Audit logging support (jansson)
* add libarchive for smbclient
* add vfs_catia module
* cleanup avahi support
* remove automatic meta package selection (avahi-daemon, krb5-server)

diizzyy commented:
IRL keeps getting in the way, it seems to work fine as far as I can tell however I'll commit to do more testing tonight.

https://github.com/openwrt/packages/pull/5563/files#diff-824425a8bd98d0e6427671b647c027fdR94
These clashes :/

Andy2244 commented:
That's fine i guess working on openwrt is quite a time-consuming side-job, i use my own feed and have no problems compiling my own FW. However i got some feedback and requests to also provide the .ipk's, which is just too much work for me.

I also hope that maybe after the initial PR is done, some more qualified person might be able to split the package apart similar to samb36, but its just to time-consuming for me atm.

PS: I also would like to add individual vfs module config support via menu, but seem like i would have to add 30 config symbols and if checks, which atm seems not worth it.

diizzyy commented:
1AM, it seems to work fine on my WRT3200ACM however I did use a custom .conf but otherwise it looks good as far as I can tell. Awesome work :D 
I'll push some more files back and forth tomorrow to see if something breaks but I doubt it.

Andy2244 commented:
@diizzyy Ok cool, if you have any additional ideas, requests i'm happy to-do everything i can, really want to finally get this thing merged.

My goal for now is to have openwrt provide a stable up-to-date smb 3 fileserver that tracks upstream, while also being easy to maintain for me. _(I'm happy to hand over the package to someone else, that want to spend more time on it!)_

hnyman commented:
I think that  utils/qemu-userspace/Makefile should not be included in the commit. 
It is unrelated to samba4, right?

what is the correlation to #4875 ?
Same stuff, or more advanced?  You also have the LuCI app ready, right?

(I am not using samba myself, so I am just wondering if this PR is ready for merge. @diizzyy finds it ok, so that tells me that the packages works.)

hnyman commented:
> Still not sure how i want to handle the luci package for it, either a complete separate one, which is what i use atm in my external feed or add some if checks to the current samba36 compatible version, which adds the extra stuff for samba4.

complete separate app, so that it can define the dependency for samba4. Then the end-user can just select  luci-app-samba4 and also samba4 gets auto-selected.

Andy2244 commented:
@hnyman Hi, utils/qemu-userspace/Makefile is actually whats needed to cross compile this package, i made a version specifically to handle host compilation of this type. This could also be used to actually get "correct" versions of tevent, tdb. This whole cross-answer file stuff is just a horrible idea in my opinion, that's why i don't use it.

#4875 was actually what inspired me to make this PR, after months of no progress and not being any closer to a merge i had enough and started my own PR. 

Main difference to #4875:
* i use qemu to cross compile, instead of cross-answer files
* i only include a minimum of patches (i'm not the code nanny for the samba project)
* is based on krb5-libs, instead of embedded heimdal (was really hard to cross compile and seems what is preferred for AD-DC usage)
* has more config options VFS modules, AD-DC, AVAHI

PS: Regarding luci, ok this makes sense will prepare a PR, if this one is ready for merge. Did not change much, my feed with it is here: https://github.com/Andy2244/openwrt-extra

Andy2244 commented:
* update to 4.8.3
* removed two patches, fixed by upstream

Andy2244 commented:
@diizzyy @lucize Hi, i found a example on how a build would look like if you really wanted to sort out the libs per binary/config selection. The only distro that seem to-do this atm is alpine: [samba4](https://git.alpinelinux.org/cgit/aports/tree/main/samba/APKBUILD)

As you can see you would need to reverse engineer the lib deps per binary or config selection and sort out unique vs shared libs. My main concern is that adapting this, also results in a kinda brittle package, which means we would have to update/test changed lib deps, names on each samba upstream update for all config permutations or write a really solid script that does this.

If anyone wants to adapt this, as noted i'm happy to add maintainers :)

lucize commented:
I'm a bit busy with ath79 targets ATM, but I'll see if I can switch to something else, just for relaxing 😄

Andy2244 commented:
@lucize No worry's, i will actually try this crazy alpine approach and try break the package apart. The main problem is that admin/client tools can't be separately installed atm, which should be at least possible. I will see how far i get and if this approach can result in a stable package.

diizzyy commented:
I agree that it's most likely worth the hassle however I don't think it's the end of the world if you build some binaries etc and don't install them. It's not like a fully fledged Samba is needed anyway. :-)

Andy2244 commented:
* refactor into server/client/admin/utils and /libs packages
* switch to native waf tools, instead of buggy make wrappers
* libreadline unbundle patch

@diizzyy ok did refactor the makefile so it behaves similar to the samab36 packages, also squashed some other minor bugs. I did spend a day trying to sort out the libs similar to the alpine makefile, but its just to darn time consuming. It seems that nearly every binary in samba4 pulls in 70-90% of the full server daemons shared libs, so from a size standpoint its pretty useless.
So tell me if you think this iteration is ready to consider to-be merged or what you would like to have changed.

Andy2244 commented:
* cleanup lib dependencies

diizzyy commented:
Awesome work! I'll try to squeeze in some testing during the weekend.

xabolcs commented:
**Offtopic**
Am I see correctly that this PR helps a lot in porting `waf` based programs to OpenWrt?

Andy2244 commented:
@xabolcs I'm only aware of the samba.org waf based programs and for those you could adapt the separate lib packages (tdb, ldb, tevent ...) to `cross-execute` instead of `cross-answer`, other than that it can be an example i guess.

xabolcs commented:
I would like to give a try to [xmms2](https://github.com/xmms2/wiki/wiki/Install-instructions).
Sure I meant using as an example, with that new `qemu-userspace` package.

Andy2244 commented:
trivial update:
* use $(PKG_JOBS)
* give credit to other devs

diizzyy commented:
...and it runs great! =)
I'm terribly sorry for the slow reply :(

I think the "old" ```smbpasswd backend``` should be default primarily for size reasons.

```
-rw-------    1 root     root      412.0K Jul 14 11:17 passdb.tdb
-rw-------    1 root     root      420.0K Jul 19 11:26 secrets.tdb
-rw-------    1 root     root         101 Jul 19 11:26 smbpasswd
```

This should also be added as default to avoid cluttering

```
veto files = /Thumbs.db/.DS_Store/._.DS_Store/.apdisk/
delete veto files = yes
```

Since even Microsoft is starting to phase it out, we should also disable NETBIOS by default. This is case less headache further down the road.

At some point we should probably have a look at LTO to further downside the library (if possible) as it's a bit of a struggle even on 16Mbyte flash to fit Samba and reasonable set of packages (btrfs eats a lot of space for instance). LTO right now it's doesn't even compile however I'm not too fond of having to maintain lots of patches for Samba as it tends to be more hassle than its worth. :-/

Tested on: WiTi board, OpenWrt Master (custom smb4.conf)

Andy2244 commented:
> Since even Microsoft is starting to phase it out, we should also disable NETBIOS by default. This is case less headache further down the road.

"disable" as in build without nmbd binary or just disable via init script, because of the samba lib snafu removing the nmbd binary saves like 100kb or so.

Will add the config suggestions to the template, any other major differences from the default template.conf compared to your custom smb4.conf?

diizzyy commented:
Not really, just a few minor differences

```
deadtime = 15 (30 seems a bit excessive)
server services = s3fs, +smb 
```
Not sure if messing with ```server service``` is needed, Samba seems to figure out this on its own? It's probably also going to be a pain setup as we have several different variants.

I don't set ```socket options```, can't really see a difference in performance and fewer lines are better ;-)

I'd suggest to remove netbios completely by default. In my case disabling netbios also implies ``` disable netbios = yes ``` in smbd.conf.

Andy2244 commented:
@diizzyy yeah i fiddled around with `server services` and its not really worth the trouble, samba does a good job starting whats needed based on security level and config options.

`socket options` yeah i'm using those options for so long, that i must admit i never tested without them, only against the WAN suggested options. As far as i'm aware they are still recommend by samba4 guides and should not hurt in LAN use cases.

**Off topic**: I still look for a way to announce samba resources to Windows Clients, to replace netbios, yet the only "simple" way would be a Windows compatible WS-Discovery service. Yet i could not find a decent working one for linux. All i found was this: [python-ws-discovery](https://github.com/andreikop/python-ws-discovery).
Are you aware of any working way to announce the samba shares to Windows 10, similar to the old netbios ways aka for Workgroups not Domains?

diizzyy commented:
https://bugzilla.samba.org/show_bug.cgi?id=11473 - Seems to be stuck... :-/

https://github.com/christgau/wsdd
Not really a fan of having python as a dependency however

https://github.com/kochinc/wsdd2 looks much more promising, might be worth looking at upstream if there are updates before importing however.

Andy2244 commented:
@diizzyy cool thanks, seems some more stuff has poped up in the last 9 months, will try this wsdd2 first.

diizzyy commented:
@Andy2244 

Looks like there's a newer version available in the latest ReadyNAS source code tarball
```
wsdd2 (1.8) stable; urgency=medium

  * Fix CPU spinner on disconnected endpoint socket.

 -- Justin Maggard <justin.maggard@netgear.com>  Fri, 06 Oct 2017 10:58:48 -0700

wsdd2 (1.7) stable; urgency=medium

  * Fix send return code evaluation.
```

I don't really know where the patches in the repo originates, can't find them public.

Andy2244 commented:
@diizzyy Yeah will create my own repo, since the existing seems unmaintained. Will use the latest 1.8 version and maybe i can keep maintaining it along the samba package.

diizzyy commented:
Managed to get LTO building by adding -flto -fPIC to both CFLAGS and LDFLAGS. Unfortunately it's only ~50kbyte ipk size with -O2 optimization so I'd say it's not worth the hassle.

Andy2244 commented:
@diizzyy I created a wsdd2 package and it worked fine on my x86_64 VM and the share was visible after starting via `wsdd2 -4 -w`, yet on openwrt nothing happens, not even a error. The VM has no firewall, so i suspect that's the problem.

I only have basic linux firewall knowledge, so maybe you can check/play around with the package (https://github.com/Andy2244/openwrt-extra)? 
I found this regarding firewall/ports: https://github.com/christgau/wsdd/blob/master/README.md

diizzyy commented:
I'm not very good at Linux network routing/firewalling either :-/
I'm seeing some multicast traffic on my Windows client but it seems incomplete
@pprindeville 
Any ideas?

lucize commented:
doing`service firewall stop` doesn't help ? to single-out the firewall

diizzyy commented:
@lucize 
I'd love to try that but my testbox is somewhat busy right now so I can't kill the fw :-/
I'll see if I can get another testbox available

Andy2244 commented:
@lucize ah you are right, did a quick test and did not help, than i noticed there is a actual error in the syslog...
`user.err wsdd2[12526]: error: wsdd-mcast-v4: wsd_init: uuid_endpoint`
The daemon needs a /etc/machine-id, which was present on my ClearLinux VM. After creating one on openwrt the daemon actually works and all my samba4 shares show up now!

Whats up with this machine-id? http://man7.org/linux/man-pages/man5/machine-id.5.html
Do i patch the source or do we create one? Ideas?

diizzyy commented:
Perhaps we can patch it to use https://github.com/openwrt/packages/blob/master/net/miniupnpd/files/miniupnpd.init#L165 instead?

Andy2244 commented:
@diizzyy Yeah i think i will just create the `/etc/machine-id` file via: `/proc/sys/kernel/random/uuid` on service start, if it does not already exists or is this a problem? I want to avoid patching for such a minor change.

diizzyy commented:
I personally don't mind but I don't know the official stance on sprinkling systemd files over the place.

Andy2244 commented:
@diizzyy Good point, found a better candidate for openwrt: `/proc/sys/kernel/random/boot_id`
Will patch the source to use this if `machine-id` is not found.

Andy2244 commented:
* replace netbios default with wsdd2 daemon
* switch to old smbpasswd backend
* add 'veto files' to smb.conf template
* revert parallel build change

@diizzyy ok, added your suggested changes and switched from netbios to wsdd2 by default. Give it a try and if you are happy, maybe we can merge this before the 18.06 release?
Will create the corresponding luci-ui PR on the weekend.

diizzyy commented:
@Andy2244 
You got mail

Andy2244 commented:
* cleanup/adhere to openwrt guidelines, fix grammar

@diizzyy No worries, i'm just glad someone finally had the time reviewing it and all changes applied.

diizzyy commented:
@Andy2244 
It's way beyond what I would be capable to accomplish so I can just offer the little knowledge I have.

Yeah, Travis doesn't seem to be happy.

@hnyman @karlp 
Can you have a look at this?

Andy2244 commented:
trivial update
* fix dir / endings
* update wsdd2

Andy2244 commented:
* update qemu-userspace to 3.0.0-rc2

Andy2244 commented:
@diizzyy So whats left for me to-do, to get this merged? Do i need to rally, find additional reviewers?

jow- commented:
I'll see no reason to not merge it, so here we go :)

Andy2244 commented:
@jow- cool thanks, seems my 6 month odyssey is finally over, now i can just hope all the testing was worth it and hopefully this fails only on a few targets :)

val-kulkov commented:
Some initial feedback without much further testing -- building samba4 with all options enabled on bcm53xx fails due to missing dependencies:

```
Package samba4-server is missing dependencies for the following libraries:
libgcrypt.so.20
libpam.so.0
Makefile:381: recipe for target '<snip>/source/bin/packages/arm_cortex-a9/packages/samba4-server_4.8.3-1_arm_cortex-a9.ipk' failed
```

The libpam dependency looks odd, given the "--without-pam" parameter in CONFIGURE_ARGS.

Andy2244 commented:
@val-kulkov 
yeah `001-samba-4.4.0-pam.patch` and `--without-pam` should take care of pam, yet samba4 likes to pull in everything it finds...

`libgcrypt.so.20` mhh never had this missing lib on my ubuntu/clear linux hosts, have to check where/when it pulls this in. What host distro do you use, maybe i can reproduce this?

val-kulkov commented:
More on the same topic:
```
Package samba4-libs is missing dependencies for the following libraries:
liblber-2.4.so.2
libldap-2.4.so.2
Makefile:380: recipe for target '/home/val/Development/lede/source/bin/packages/arm_cortex-a9/packages/samba4-libs_4.8.3-1_arm_cortex-a9.ipk' failed
```
My host is Ubuntu 16.04. The target, again, is bcm53xx.

Andy2244 commented:
Thanks, this seems strange the ldap libs should come from `+SAMBA4_SERVER_AD_DC:libopenldap`. Will try to reproduce this.

Andy2244 commented:
oki, will also try with `CONFIG_PACKAGE_libpam=y`, regarding the other missing libs, need to recompile for your target `bcm53xx` locally arm/mips still work with all options enabled.

diizzyy commented:
@Andy2244 
Usually you can't really test for these dependency issues due to race conditions unfortunately unless you happen to have a full package repo from the start. Samba is a major pain when it comes to dependencies as it tends to pull in whatever it finds even if it's not needed as you've already noticed. :/

val-kulkov commented:
The addition of "+PACKAGE_libgcrypt:libgcrypt +PACKAGE_libpam:libpam" to line 52 in Makefile solved the issue for me. All five samba4 packages compiled successfully. Mysteriously, there were no complaints about liblber or libldap this time.

diizzyy commented:
@pprindeville 
It's all ok, Andy managed to figure it out :-)
